drivers/spi/spi_nrf5.c

```diff
@@ -0,0 +1,481 @@
+/*
+ * Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "spi.h"
+#include "nrfx_spi.h"
+
+#define SYS_LOG_DOMAIN "SPI NRF5"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_SPI_LEVEL
+#include <logging/sys_log.h>
+
+#include "spi_context.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define SPI_NRF5_MAX_SLAVES_ON_BUS (4)
+
+typedef void (*irq_cfg_func_t)(void);
+
+/** Configuration data */
+struct spi_nrf5_cfg {
+	nrfx_spi_t spi;
+	irq_cfg_func_t irq_cfg_func;
+	struct {
+		u8_t sck;
+		u8_t mosi;
+		u8_t miso;
+		/* Pin numbers of up to 4 slave devices */
+		u8_t ss[SPI_NRF5_MAX_SLAVES_ON_BUS];
+	} psel;
+	u8_t orc;
+};
+
+/** Driver data */
+struct spi_nrf5_data {
+	struct spi_context ctx;
+	u16_t transfer_len;
+	bool busy;
+};
+
+#define DEV_CFG(spi_cfg) ((const struct spi_nrf5_cfg * const) \
+					(spi_cfg->dev)->config->config_info)
+#define DEV_DATA(spi_cfg) ((struct spi_nrf5_data * const) \
+					(spi_cfg->dev)->driver_data)
+
+static inline nrfx_spi_frequency_t spi_freq_to_nrfx_freq(u32_t frequency)
+{
+	switch (frequency) {
+	case 125000:
+		return NRFX_SPI_FREQ_125K;
+	case 250000:
+		return NRFX_SPI_FREQ_250K;
+	case 500000:
+		return NRFX_SPI_FREQ_500K;
+	case 1000000:
+		return NRFX_SPI_FREQ_1M;
+	case 2000000:
+		return NRFX_SPI_FREQ_2M;
+	case 4000000:
+		return NRFX_SPI_FREQ_4M;
+	case 8000000:
+		return NRFX_SPI_FREQ_8M;
+	default:
+		SYS_LOG_WRN("Unsupported frequency, setting 125 kbps");
+		return NRFX_SPI_FREQ_125K;
+	}
+}
+
+static inline nrfx_spi_mode_t spi_operation_to_nrfx_mode(u16_t operation)
+{
+	if (SPI_MODE_GET(operation) & SPI_MODE_CPOL) {
+		if (SPI_MODE_GET(operation) & SPI_MODE_CPHA) {
+			return NRFX_SPI_MODE_3;
+		} else {
+			return NRFX_SPI_MODE_2;
+		}
+	} else {
+		if (SPI_MODE_GET(operation) & SPI_MODE_CPHA) {
+			return NRFX_SPI_MODE_1;
+		} else {
+			return NRFX_SPI_MODE_0;
+		}
+	}
+}
+
+static inline
+nrfx_spi_bit_order_t spi_operation_to_nrfx_bit_order(u16_t operation)
+{
+	if (operation & SPI_TRANSFER_LSB) {
+		return NRFX_SPI_BIT_ORDER_LSB_FIRST;
+	} else {
+		return NRFX_SPI_BIT_ORDER_MSB_FIRST;
+	}
+}
+
+static inline int set_nrfx_spi_config(nrfx_spi_config_t *nrfx_spi_cfg,
+				      const struct spi_config *spi_cfg)
+{
+	const struct spi_nrf5_cfg *dev_cfg = DEV_CFG(spi_cfg);
+	u8_t slave = spi_cfg->slave;
+
+	nrfx_spi_cfg->sck_pin  = dev_cfg->psel.sck;
```

Comment:
The GPIO pins for the SPI should only be configured once during init on not per configuration change